python/redis_work_queue/workqueue.py

```diff
@@ -37,6 +36,41 @@ def add_item(self, db: Redis, item: Item) -> None:
         pipeline = db.pipeline()
         self.add_item_to_pipeline(pipeline, item)
         pipeline.execute()
+    
+    def add_new_item(self, db: Redis, item: Item) -> bool:
+        """Adds an item to the work queue only if an item with the same ID doesn't already exist.
+
+        This method uses WATCH to add the item atomically. The db client passed must not be used by
+        anything else while this method is running.
+
+        Returns a boolean indicating if the item was added or not. The item is only not added if it
+        already exists or if an error (other than a transaction error, which triggers a retry) occurs."""
+
+        while True:
+            try:
+                pipeline = db.pipeline(transaction=True)
+                pipeline.watch(self._main_queue_key, self._processing_key)
+
+                if (
+                    pipeline.lpos(self._main_queue_key, item.id()) is not None
+                    or pipeline.lpos(self._processing_key, item.id()) is not None
+                ):
+                    pipeline.unwatch()
+                    return False
+
+                pipeline.multi()
+
+                self.add_item_to_pipeline(pipeline, item)
+
+                pipeline.execute()
+                return True
+
+            except redis.WatchError:
+                continue
+            except Exception as e:
+                print("Error:", e)
+                raise
```

Comment:
We don't need to print the error if we're raising it.

In fact, do we actually need this final `except`?